File: HighFive/HighFive/System/Text/UTF7Encoding.cs
namespace System.Text
{
    public class UTF7Encoding : Encoding
    {
        private bool allowOptionals;

        public UTF7Encoding() : this(false)
        {
        }

        public UTF7Encoding(bool allowOptionals)
        {
            this.allowOptionals = allowOptionals;
            this.fallbackCharacter = '\uFFFD';
        }

        public override int CodePage => 65000;
        public override string EncodingName => "Unicode (UTF-7)";

        private static string Escape(string chars)
        {
            return HighFive.Script.Write<string>("chars.replace(/[-[\\]{}()*+?.,\\\\^$|#\\s]/g, \"\\\\$&\")");
        }

        protected override byte[] Encode(string s, byte[] outputBytes, int outputIndex, out int writtenBytes)
        {
            var setD = "A-Za-z0-9" + Escape("'(),-./:?");

            Func<string, string> encode = (str) =>
            {
                var b = new byte[str.Length * 2];
                var bi = 0;
                for (var i = 0; i < str.Length; i++)
                {
                    var c = str[i];
                    b[bi++] = (c >> 8).As<byte>();
                    b[bi++] = (c & 0xFF).As<byte>();
                }
                var base64Str = System.Convert.ToBase64String(b);
                return HighFive.Script.Write<string>("base64Str.replace(/=+$/, \"\")");
            };

            var setO = Escape("!\"#$%&*;<=>@[]^_`{|}");
            var setW = Escape(" \r\n\t");

            s = HighFive.Script.Write<string>("s.replace(new RegExp(\"[^\" + setW + setD + (this.allowOptionals ? setO : \"\") + \"]+\", \"g\"), function (chunk) { return \"+\" + (chunk === \"+\" ? \"\" : encode(chunk)) + \"-\"; })");

            var arr = s.ToCharArray();

            if (outputBytes != null)
            {
                var recorded = 0;

                if (arr.Length > (outputBytes.Length - outputIndex))
                {
                    throw new System.ArgumentException("bytes");
                }

                for (var j = 0; j < arr.Length; j++)
                {
                    outputBytes[j + outputIndex] = arr[j].As<byte>();
                    recorded++;
                }

                writtenBytes = recorded;
                return null;
            }

            writtenBytes = arr.Length;

            return arr.As<byte[]>();
        }

        protected override string Decode(byte[] bytes, int index, int count, char[] chars, int charIndex)
        {
            Func<string, char[]> _base64ToArrayBuffer = (base64) =>
            {
                try
                {
                    HighFive.Script.Write("if (typeof window === \"undefined\") { throw new System.Exception(); }");
                    var binary_string = HighFive.Script.Write<string>("window.atob(base64)");
                    var len = binary_string.Length;
                    var arr = new char[len];

                    if (len == 1 && binary_string[0] == 0)
                    {
                        return new char[0];
                    }

                    for (var i = 0; i < len; i++)
                    {
                        arr[i] = binary_string[i];
                    }

                    return arr;
                }
                catch (Exception)
                {
                    return new char[0];
                }
            };

            Func<string, string> decode = (s) =>
            {
                var b = _base64ToArrayBuffer(s);
                var r = new char[0];
                for (var i = 0; i < b.Length;)
                {
                    r.Push((char)(b[i++] << 8 | b[i++]));
                }
                return new string(r);
            };

            var str = new string(bytes.As<char[]>(), index, count);
            return HighFive.Script.Write<string>(@"str.replace(/\+([A-Za-z0-9\/]*)-?/gi, function (_, chunk) { if (chunk === """") { return _ == ""+-"" ? ""+"" : """"; } return decode(chunk); })");
        }

        public override int GetMaxByteCount(int charCount)
        {
            if (charCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }

            var byteCount = (long)charCount * 3 + 2;

            if (byteCount > 0x7fffffff)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }

            return (int)byteCount;
        }

        public override int GetMaxCharCount(int byteCount)
        {
            if (byteCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("byteCount");
            }

            var charCount = byteCount;
            if (charCount == 0)
            {
                charCount = 1;
            }

            return charCount | 0;
        }
    }
}

File: HighFive/HighFive/Resources/.generated/System/Text/UTF7Encoding.js
    HighFive.define("System.Text.UTF7Encoding", {
        inherits: [System.Text.Encoding],
        statics: {
            methods: {
                Escape: function (chars) {
                    return chars.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
                }
            }
        },
        fields: {
            allowOptionals: false
        },
        props: {
            CodePage: {
                get: function () {
                    return 65000;
                }
            },
            EncodingName: {
                get: function () {
                    return "Unicode (UTF-7)";
                }
            }
        },
        ctors: {
            ctor: function () {
                System.Text.UTF7Encoding.$ctor1.call(this, false);
            },
            $ctor1: function (allowOptionals) {
                this.$initialize();
                System.Text.Encoding.ctor.call(this);
                this.allowOptionals = allowOptionals;
                this.fallbackCharacter = 65533;
            }
        },
        methods: {
            Encode$3: function (s, outputBytes, outputIndex, writtenBytes) {
                var setD = "A-Za-z0-9" + (System.Text.UTF7Encoding.Escape("'(),-./:?") || "");

                var encode = $asm.$.System.Text.UTF7Encoding.f1;

                var setO = System.Text.UTF7Encoding.Escape("!\"#$%&*;<=>@[]^_`{|}");
                var setW = System.Text.UTF7Encoding.Escape(" \r\n\t");

                s = s.replace(new RegExp("[^" + setW + setD + (this.allowOptionals ? setO : "") + "]+", "g"), function (chunk) { return "+" + (chunk === "+" ? "" : encode(chunk)) + "-"; });

                var arr = System.String.toCharArray(s, 0, s.length);

                if (outputBytes != null) {
                    var recorded = 0;

                    if (arr.length > (((outputBytes.length - outputIndex) | 0))) {
                        throw new System.ArgumentException.$ctor1("bytes");
                    }

                    for (var j = 0; j < arr.length; j = (j + 1) | 0) {
                        outputBytes[System.Array.index(((j + outputIndex) | 0), outputBytes)] = arr[System.Array.index(j, arr)];
                        recorded = (recorded + 1) | 0;
                    }

                    writtenBytes.v = recorded;
                    return null;
                }

                writtenBytes.v = arr.length;

                return arr;
            },
            Decode$2: function (bytes, index, count, chars, charIndex) {
                var _base64ToArrayBuffer = $asm.$.System.Text.UTF7Encoding.f2;

                var decode = function (s) {
                    var b = _base64ToArrayBuffer(s);
                    var r = System.Array.init(0, 0, System.Char);
                    for (var i = 0; i < b.length; ) {
                        r.push(((b[System.Array.index(HighFive.identity(i, ((i = (i + 1) | 0))), b)] << 8 | b[System.Array.index(HighFive.identity(i, ((i = (i + 1) | 0))), b)]) & 65535));
                    }
                    return System.String.fromCharArray(r);
                };

                var str = System.String.fromCharArray(bytes, index, count);
                return str.replace(/\+([A-Za-z0-9\/]*)-?/gi, function (_, chunk) { if (chunk === "") { return _ == "+-" ? "+" : ""; } return decode(chunk); });
            },
            GetMaxByteCount: function (charCount) {
                if (charCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }

                var byteCount = System.Int64(charCount).mul(System.Int64(3)).add(System.Int64(2));

                if (byteCount.gt(System.Int64(2147483647))) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }

                return System.Int64.clip32(byteCount);
            },
            GetMaxCharCount: function (byteCount) {
                if (byteCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("byteCount");
                }

                var charCount = byteCount;
                if (charCount === 0) {
                    charCount = 1;
                }

                return charCount | 0;
            }
        }
    });

    HighFive.ns("System.Text.UTF7Encoding", $asm.$);

    HighFive.apply($asm.$.System.Text.UTF7Encoding, {
        f1: function (str) {
            var b = System.Array.init(HighFive.Int.mul(str.length, 2), 0, System.Byte);
            var bi = 0;
            for (var i = 0; i < str.length; i = (i + 1) | 0) {
                var c = str.charCodeAt(i);
                b[System.Array.index(HighFive.identity(bi, ((bi = (bi + 1) | 0))), b)] = (c >> 8);
                b[System.Array.index(HighFive.identity(bi, ((bi = (bi + 1) | 0))), b)] = (c & 255);
            }
            var base64Str = System.Convert.toBase64String(b, null, null, null);
            return base64Str.replace(/=+$/, "");
        },
        f2: function (base64) {
            try {
                if (typeof window === "undefined") { throw new System.Exception(); };
                var binary_string = window.atob(base64);
                var len = binary_string.length;
                var arr = System.Array.init(len, 0, System.Char);

                if (len === 1 && binary_string.charCodeAt(0) === 0) {
                    return System.Array.init(0, 0, System.Char);
                }

                for (var i = 0; i < len; i = (i + 1) | 0) {
                    arr[System.Array.index(i, arr)] = binary_string.charCodeAt(i);
                }

                return arr;
            } catch ($e1) {
                $e1 = System.Exception.create($e1);
                return System.Array.init(0, 0, System.Char);
            }
        }
    });
